File: R/get_met.R
################################################################################
# Code started by Molly Stroud on 12/16/25
# Download data from dynamical.org and put into correct formatting for FLARE
# https://dynamical.org/catalog/noaa-gefs-forecast-35-day/
#
# This used to drive Python via reticulate, which broke on other machines due
# to reticulate/virtualenv version and configuration mismatches. All of the
# xarray/zarr/pandas logic now lives in R/get_met.py (pure Python, no
# reticulate). This file just provides helpers to set up a dedicated Python
# virtual environment and run get_met.py as a subprocess; 02Get_Inputs.R calls
# .run_get_met_py() directly for stage 2 / stage 3 downloads.
################################################################################

.get_project_root <- function() {
  root <- tryCatch(here::here(), error = function(e) NA_character_)
  if (!is.na(root) && nzchar(root) && dir.exists(root)) {
    return(normalizePath(root, winslash = "/", mustWork = FALSE))
  }
  normalizePath(getwd(), winslash = "/", mustWork = FALSE)
}

.get_met_py_script <- file.path(.get_project_root(), "R", "get_met.py")
.get_met_venv <- file.path(.get_project_root(), ".venv-met")

.normalize_stage3_met <- function(site_id, bbox, base_dir = file.path("drivers", "met", "gefs-v12", "stage3")) {
  met_dir <- file.path(base_dir, paste0("site_id=", site_id))
  met_file <- file.path(met_dir, "part-0.parquet")

  if (!file.exists(met_file)) {
    return(invisible(FALSE))
  }

  met_df <- arrow::read_parquet(met_file)
  met_df$datetime <- lubridate::as_datetime(met_df$datetime, tz = "UTC")

  if (!"site_id" %in% names(met_df)) {
    met_df$site_id <- site_id
  }

  unique_time <- sort(unique(met_df$datetime))
  needs_hourly <- length(unique_time) >= 2 &&
    as.numeric(difftime(unique_time[2], unique_time[1], units = "hours")) > 1

  flux_vars <- c("precipitation_flux", "surface_downwelling_shortwave_flux_in_air", "surface_downwelling_longwave_flux_in_air")
  flux_summary <- met_df |>
    dplyr::filter(variable %in% flux_vars) |>
    dplyr::group_by(variable) |>
    dplyr::summarise(non_na = sum(!is.na(prediction)), .groups = "drop")
  missing_flux_var <- any(!flux_vars %in% flux_summary$variable)
  empty_flux_var <- any(flux_summary$non_na == 0)
  needs_flux_repair <- missing_flux_var || empty_flux_var

  if (!needs_hourly && !needs_flux_repair) {
    return(invisible(FALSE))
  }

  source(file.path("R", "to_hourly.R"), local = TRUE)

  if (needs_hourly) {
    mean_lon <- mean(c(as.numeric(bbox[["left"]]), as.numeric(bbox[["right"]])))
    mean_lat <- mean(c(as.numeric(bbox[["bottom"]]), as.numeric(bbox[["top"]])))
    met_df <- get_hourly(met_df,
                         mean_lon = mean_lon,
                         mean_lat = mean_lat)
    met_df$datetime <- lubridate::as_datetime(met_df$datetime, tz = "UTC")
  }

  if (needs_flux_repair) {
    if (!"site_id" %in% names(met_df)) {
      met_df$site_id <- site_id
    }

    air_temp_lookup <- met_df |>
      dplyr::filter(variable == "air_temperature") |>
      dplyr::select(parameter, datetime, air_temp_k = prediction)

    mean_lon <- mean(c(as.numeric(bbox[["left"]]), as.numeric(bbox[["right"]])))
    mean_lat <- mean(c(as.numeric(bbox[["bottom"]]), as.numeric(bbox[["top"]])))
    solar_lon <- ifelse(mean_lon < 0, 360 + mean_lon, mean_lon)

    met_df <- met_df |>
      dplyr::left_join(air_temp_lookup, by = c("parameter", "datetime")) |>
      dplyr::mutate(
        prediction = dplyr::case_when(
          variable == "precipitation_flux" & is.na(prediction) ~ 0,
          variable == "surface_downwelling_shortwave_flux_in_air" & is.na(prediction) ~
            downscale_solar_geom(
              lubridate::yday(datetime) + lubridate::hour(datetime) / 24,
              solar_lon,
              mean_lat
            ) * 0.5,
          variable == "surface_downwelling_longwave_flux_in_air" & is.na(prediction) ~
            0.97 * 5.67e-8 * (air_temp_k^4),
          TRUE ~ prediction
        )
      ) |>
      dplyr::select(-air_temp_k) |>
      dplyr::group_by(site_id, family, parameter, variable) |>
      dplyr::arrange(datetime, .by_group = TRUE) |>
      tidyr::fill(prediction, .direction = "downup") |>
      dplyr::ungroup()
  }

  met_df |>
    dplyr::select(-site_id) |>
    arrow::write_parquet(met_file)

  message(paste0("Normalized stage3 met data for site ", site_id,
                 " (hourly: ", needs_hourly,
                 ", flux_repair: ", needs_flux_repair, ")."))
  invisible(TRUE)
}

# path to the python executable inside the dedicated virtual environment
.get_met_venv_python <- function() {
  if (.Platform$OS.type == "windows") {
    file.path(.get_met_venv, "Scripts", "python.exe")
  } else {
    file.path(.get_met_venv, "bin", "python3")
  }
}

# make sure a python3 interpreter + dedicated venv with the right packages exist
.find_system_python <- function() {
  if (.Platform$OS.type == "windows") {
    py_launcher <- Sys.which("py")
    if (nzchar(py_launcher)) {
      return(list(cmd = py_launcher, args = c("-3"), label = "py -3"))
    }
  }

  python3 <- Sys.which("python3")
  if (nzchar(python3)) {
    return(list(cmd = python3, args = character(0), label = "python3"))
  }

  python <- Sys.which("python")
  if (nzchar(python)) {
    return(list(cmd = python, args = character(0), label = "python"))
  }

  stop(
    "Could not find a Python 3 interpreter on PATH. Tried py -3 (Windows), ",
    "python3, and python. Please install Python 3 and try again."
  )
}

.ensure_met_python_env <- function() {
  system_python <- .find_system_python()

  venv_python <- .get_met_venv_python()

  if (!file.exists(.get_met_py_script)) {
    stop("Could not find get_met.py at: ", .get_met_py_script)
  }

  if (!file.exists(venv_python)) {
    message("Creating a dedicated Python environment for met data downloads (.venv-met)...")
    result <- system2(system_python$cmd, c(system_python$args, "-m", "venv", .get_met_venv))
    if (result != 0) {
      stop(
        "Failed to create the Python virtual environment used by get_met.py ",
        "with interpreter: ", system_python$label
      )
    }
  }

  # bump this when `packages` below changes so existing .venv-met installs
  # (which skip installation once the marker file exists) get upgraded
  required_marker <- "packages-v2"

  marker <- file.path(.get_met_venv, "packages_installed.txt")
  installed_marker <- if (file.exists(marker)) readLines(marker, warn = FALSE)[1] else NA_character_

  if (!identical(installed_marker, required_marker)) {
    message("Installing required Python packages for met data downloads...")
    packages <- c(
      "dask==2025.1.0",
      "xarray[complete]==2026.1.0",
      "zarr>=3.2,<4",
      "certifi",
      "numpy",
      "pandas",
      "pyarrow",
      "requests",
      "aiohttp",
      "dynamical-catalog"
    )
    # shQuote() protects package specifiers with shell-special characters
    # (e.g. "zarr>=3.2,<4") from being interpreted by the shell that
    # system2() invokes them through
    system2(venv_python, c("-m", "pip", "install", "--quiet", "--upgrade", "pip"))
    result <- system2(venv_python, c("-m", "pip", "install", "--quiet", shQuote(packages)))
    if (result != 0) {
      stop("Failed to install the Python packages required by get_met.py.")
    }
    writeLines(required_marker, marker)
  }

  venv_python
}

# Run get_met.py for a single reference_datetime.
#
# For "stage2", one forecast cycle (the full ~35-day horizon) for that
# reference_datetime is downloaded. The Python CLI uses --start-date==--end-date
# (both set to reference_datetime) so every forecast horizon is retained. For
# "stage3", reference_datetime is the start date and end_date is its optional
# end date; stage3 pulls from the NOAA GEFS *analysis* dataset (a single
# deterministic time series, not per-day forecast cycles), so the whole
# start_date..end_date window is fetched in one request.
.run_get_met_py <- function(stage, site_id, bbox, reference_datetime, end_date = NULL) {
  stage <- match.arg(stage, c("stage2", "stage3"))

  if (identical(stage, "stage2")) {
    end_date <- reference_datetime
    if (is.null(end_date)) {
      stop("reference_datetime is required for stage2.")
    }
  }

  args <- c(
    stage,
    "--site", site_id,
    "--bbox", bbox[["left"]], bbox[["bottom"]], bbox[["right"]], bbox[["top"]],
    "--start-date", as.character(reference_datetime),
    "--end-date", as.character(end_date)
  )

  venv_python <- .ensure_met_python_env()
  result <- system2(venv_python, c(.get_met_py_script, as.character(args)))
  if (result != 0) {
    stop("get_met.py failed. See the messages above for details.")
  }

  if (identical(stage, "stage3")) {
    .normalize_stage3_met(site_id = site_id, bbox = bbox)
  }

  invisible(result)
}

File: R/get_met.py
################################################################################
# Download meteorological forecast data from dynamical.org and format it for
# FLARE. This is a pure-Python port of the original get_met.R (which relied on
# reticulate to drive Python from R). Doing the xarray/zarr/pandas work in
# native Python avoids reticulate/virtualenv version mismatches between
# machines -- this script is called as a subprocess from R (see R/get_met.R).
#
# https://dynamical.org/catalog/noaa-gefs-forecast-35-day/
################################################################################
import argparse
import os

import certifi
import numpy as np
import pandas as pd
import xarray as xr

# variables of interest in the dynamical.org GEFS zarr store
VARS = [
    "temperature_2m",
    "relative_humidity_2m",
    "pressure_surface",
    "wind_u_10m",
    "wind_v_10m",
    "downward_long_wave_radiation_flux_surface",
    "downward_short_wave_radiation_flux_surface",
    "precipitation_surface",
]

# rename GEFS variable names to the names FLARE expects
VAR_RENAME = {
    "temperature_2m": "air_temperature",
    "relative_humidity_2m": "relative_humidity",
    "pressure_surface": "air_pressure",
    "wind_u_10m": "eastward_wind",
    "wind_v_10m": "northward_wind",
    "downward_long_wave_radiation_flux_surface": "surface_downwelling_longwave_flux_in_air",
    "downward_short_wave_radiation_flux_surface": "surface_downwelling_shortwave_flux_in_air",
    "precipitation_surface": "precipitation_flux",
}

STATE_VARS = [
    "air_pressure",
    "relative_humidity",
    "air_temperature",
    "eastward_wind",
    "northward_wind",
]

FLUX_VARS = [
    "precipitation_flux",
    "surface_downwelling_longwave_flux_in_air",
    "surface_downwelling_shortwave_flux_in_air",
]

FINAL_COLS = ["site_id", "family", "parameter", "datetime", "variable", "prediction"]


################################################################################
# solar geometry helpers (ported from R/to_hourly.R)
################################################################################
def equation_of_time(doy):
    f = np.pi / 180 * (279.5 + 0.9856 * doy)
    et = (
        -104.7 * np.sin(f)
        + 596.2 * np.sin(2 * f)
        + 4.3 * np.sin(4 * f)
        - 429.3 * np.cos(f)
        - 2 * np.cos(2 * f)
        + 19.3 * np.cos(3 * f)
    ) / 3600
    return et


def cos_solar_zenith_angle(doy, lat, lon, dt, hr):
    et = equation_of_time(doy)
    merid = np.floor(lon / 15) * 15
    if merid < 0:
        merid += 15
    lc = (lon - merid) * -4 / 60
    tz = merid / 360 * 24
    midbin = 0.5 * dt / 86400 * 24
    t0 = 12 + lc - et - tz - midbin
    h = np.pi / 12 * (hr - t0)
    dec = -23.45 * np.pi / 180 * np.cos(2 * np.pi * (doy + 10) / 365)
    cosz = np.sin(lat * np.pi / 180) * np.sin(dec) + np.cos(lat * np.pi / 180) * np.cos(dec) * np.cos(h)
    return np.where(cosz < 0, 0, cosz)


def downscale_solar_geom(doy, lon, lat):
    doy = np.asarray(doy, dtype=float)
    dt = np.median(np.diff(doy)) * 86400
    hr = (doy - np.floor(doy)) * 24
    cosz = cos_solar_zenith_angle(doy, lat, lon, dt, hr)
    return 1366 * cosz


################################################################################
# hourly downscaling (ported from get_hourly() in R/to_hourly.R)
################################################################################
def get_hourly(df, mean_lon, mean_lat):
    base = df[FINAL_COLS].copy()
    # GEFS values come back as float32; upcast so downstream calculations
    # (e.g. shortwave downscaling) don't overflow/raise on assignment
    base["prediction"] = base["prediction"].astype("float64")

    parameters = base["parameter"].unique()
    variables = base["variable"].unique()
    sites = base["site_id"].unique()
    datetime_range = pd.date_range(base["datetime"].min(), base["datetime"].max(), freq="1h")

    parameter_maxtime = (
        base.groupby(["site_id", "family", "parameter"])["datetime"]
        .max()
        .reset_index()
        .rename(columns={"datetime": "max_time"})
    )

    full_time = pd.MultiIndex.from_product(
        [sites, parameters, datetime_range, variables],
        names=["site_id", "parameter", "datetime", "variable"],
    ).to_frame(index=False)

    full_time = full_time.merge(parameter_maxtime, on=["site_id", "parameter"], how="left")
    full_time = full_time[full_time["datetime"] <= full_time["max_time"]]
    full_time = full_time.drop(columns=["max_time"]).drop_duplicates()

    join_keys = ["site_id", "parameter", "datetime", "family", "variable"]

    # states: linearly interpolated over time within each site/parameter/variable
    states = full_time.merge(base, on=join_keys, how="left")
    states = states[states["variable"].isin(STATE_VARS)].sort_values(["site_id", "parameter", "datetime"])
    states["prediction"] = states.groupby(["site_id", "parameter", "variable"])["prediction"].transform(
        lambda s: s.interpolate(method="linear", limit_direction="both")
    )
    states.loc[states["variable"] == "air_temperature", "prediction"] += 273
    # NOTE: raw GEFS temperature_2m from data.dynamical.org is in Celsius, so
    # the +273 converts to Kelvin (exactly once) for FLARE, which expects Kelvin.
    # The original R code checked `variable == "RH"`, a name that never
    # occurs after the rename above, so relative humidity is never divided by
    # 100 here. Preserved as-is for output parity with the legacy script.

    # fluxes: back-filled, then shortwave is downscaled using solar geometry
    fluxes = full_time.merge(base, on=join_keys, how="left")
    fluxes = fluxes[fluxes["variable"].isin(FLUX_VARS)].sort_values(["site_id", "family", "parameter", "datetime"])
    fluxes["prediction"] = fluxes.groupby(["site_id", "family", "parameter", "variable"])["prediction"].transform(
        lambda s: s.bfill()
    )
    fluxes.loc[fluxes["variable"] == "precipitation_flux", "prediction"] /= 6 * 60 * 60

    fluxes["hour"] = fluxes["datetime"].dt.hour
    fluxes["date"] = fluxes["datetime"].dt.date
    fluxes["doy"] = fluxes["datetime"].dt.dayofyear + fluxes["hour"] / 24
    fluxes["rpot"] = downscale_solar_geom(fluxes["doy"].to_numpy(), mean_lon, mean_lat)

    group_cols = ["site_id", "family", "parameter", "date", "variable"]
    fluxes["avg_sw"] = fluxes.groupby(group_cols)["prediction"].transform("mean")
    fluxes["avg_rpot"] = fluxes.groupby(group_cols)["rpot"].transform("mean")

    sw_mask = (fluxes["variable"] == "surface_downwelling_shortwave_flux_in_air") & (fluxes["avg_rpot"] > 0.0)
    fluxes.loc[sw_mask, "prediction"] = (
        fluxes.loc[sw_mask, "rpot"] * (fluxes.loc[sw_mask, "avg_sw"] / fluxes.loc[sw_mask, "avg_rpot"])
    )

    hourly_df = pd.concat([states[FINAL_COLS], fluxes[FINAL_COLS]], ignore_index=True)
    hourly_df = hourly_df.sort_values(["site_id", "family", "variable", "datetime"]).reset_index(drop=True)
    return hourly_df


################################################################################
# fetch + reformat one forecast cycle worth of met data
################################################################################
def get_temp_gefs(ds, site_id, start_time, bbox, lead_time=True, keep_only_day=True):
    lon_min, lat_min, lon_max, lat_max = bbox
    mean_lat = (lat_min + lat_max) / 2
    mean_lon = (lon_min + lon_max) / 2

    temp = ds[VARS].sel(
        init_time=str(start_time),
        latitude=mean_lat,
        longitude=mean_lon,
        method="nearest",
    )
    temp_r = temp.assign_coords(
        lead_hours=temp.lead_time.astype("timedelta64[h]").astype(int),
        member_id=temp.ensemble_member,
        init_time=temp.init_time,
    )

    if lead_time is not True:
        # NOTE: mirrors the original R logic (`lead_hours <= 86400`). Since
        # lead_hours is expressed in hours (max ~840 for a 35-day forecast)
        # this comparison never actually filters anything; preserved as-is
        # for output parity. The date filter at the end of this function is
        # what actually restricts stage 3 downloads to a single day.
        temp_r = temp_r.sel(lead_hours=temp_r.lead_hours <= 86400)

    df = temp_r.to_dataframe().reset_index()
    df = df.drop(
        columns=["expected_forecast_length", "ingested_forecast_length", "latitude", "longitude", "spatial_ref", "lead_hours"],
        errors="ignore",
    )

    id_vars = [c for c in df.columns if c not in VARS]
    temp_df = df.melt(id_vars=id_vars, value_vars=VARS, var_name="variable", value_name="prediction")
    temp_df["family"] = "ensemble"
    temp_df["site_id"] = site_id
    temp_df["init_time"] = pd.to_datetime(temp_df["init_time"]).dt.date
    temp_df = temp_df.rename(
        columns={"init_time": "reference_datetime", "valid_time": "datetime", "member_id": "parameter"}
    )
    temp_df["variable"] = temp_df["variable"].replace(VAR_RENAME)
    temp_df["datetime"] = pd.to_datetime(temp_df["datetime"], utc=True)
    temp_df["parameter"] = temp_df["parameter"].astype("float64")

    hourly_df = get_hourly(temp_df, mean_lon, mean_lat)
    if keep_only_day:
        hourly_df = hourly_df[hourly_df["datetime"].dt.date == pd.to_datetime(start_time).date()]
    return hourly_df


################################################################################
# stage 2: one parquet file per reference_datetime/site_id
################################################################################
def get_stage_2(ds, start_date, end_date, site, bbox, base_dir="drivers/met/gefs-v12/stage2"):
    dates = pd.date_range(start_date, end_date, freq="1D").date
    for date in dates:
        print(f"Downloading stage 2 met data for {date}", flush=True)
        # keep the full forecast lead time (the entire ~35-day GEFS cycle) so a
        # forecast launched at `date` is covered across its full horizon
        metdata = get_temp_gefs(ds, site, str(date), bbox, lead_time=True, keep_only_day=False).copy()
        metdata["reference_datetime"] = date
        for (ref_dt, site_id), group in metdata.groupby(["reference_datetime", "site_id"]):
            out_dir = os.path.join(base_dir, f"reference_datetime={ref_dt}", f"site_id={site_id}")
            os.makedirs(out_dir, exist_ok=True)
            group.drop(columns=["reference_datetime", "site_id"]).to_parquet(
                os.path.join(out_dir, "part-0.parquet"), index=False
            )
    print("Stage 2 data downloaded!")


################################################################################
# stage 3: historical met from start_date to end_date, one parquet file per site_id
################################################################################
def get_stage_3(ds, start_date, site, bbox, end_date=None, base_dir="drivers/met/gefs-v12/stage3"):
    start = pd.to_datetime(start_date).date()
    if end_date is None:
        end = start + pd.Timedelta(days=5)
    else:
        end = pd.to_datetime(end_date).date()
    dates = pd.date_range(start, end, freq="1D").date
    stage3 = []
    for date in dates:
        print(f"Downloading stage 3 met data for {date}", flush=True)
        metdata = get_temp_gefs(ds, site, str(date), bbox, lead_time=0)
        stage3.append(metdata)
    stage3 = pd.concat(stage3, ignore_index=True)
    for site_id, group in stage3.groupby("site_id"):
        out_dir = os.path.join(base_dir, f"site_id={site_id}")
        os.makedirs(out_dir, exist_ok=True)
        group.drop(columns=["site_id"]).to_parquet(os.path.join(out_dir, "part-0.parquet"), index=False)
    print("Stage 3 data downloaded!")


def main():
    parser = argparse.ArgumentParser(description="Download NOAA GEFS met forecasts for FLARE from data.dynamical.org")
    parser.add_argument("stage", choices=["stage2", "stage3"])
    parser.add_argument("--site", required=True, help="site_id")
    parser.add_argument(
        "--bbox", required=True, nargs=4, type=float, metavar=("LEFT", "BOTTOM", "RIGHT", "TOP")
    )
    parser.add_argument("--start-date", required=True)
    parser.add_argument("--end-date", required=False, help="required for stage2; optional for stage3 (defaults to start+5 days)")
    args = parser.parse_args()

    # make sure http(s) access works regardless of the machine's default certs
    os.environ["SSL_CERT_FILE"] = certifi.where()

    print("Opening data from data.dynamical.org")
    ds = xr.open_zarr(
        "https://data.dynamical.org/noaa/gefs/forecast-35-day/latest.zarr?email=optional@email.com",
        consolidated=True,
        decode_timedelta=False,
        chunks="auto",
    )

    if args.stage == "stage2":
        if not args.end_date:
            parser.error("--end-date is required for stage2")
        get_stage_2(ds, args.start_date, args.end_date, args.site, args.bbox)
    else:
        get_stage_3(ds, args.start_date, args.site, args.bbox, end_date=args.end_date)


if __name__ == "__main__":
    main()
